tree.go

```diff
@@ -820,24 +725,33 @@ func (n *InternalNode) Serialize() ([]byte, error) {
 			}
 		}
 	}
-	children := make([]byte, 0, (len(commitments)+nhashed)*32)
 
-	bytecomms := banderwagon.ElementsToBytes(commitments)
+	ret := make([]byte, nodeTypeSize+bitlistSize+(len(commitments)+nhashed)*SerializedPointSize)
+
+	// We create a children slice from ret ready to start appending children without allocations.
+	children := ret[internalNodeChildrenOffset:internalNodeChildrenOffset]
+	bytecomms := banderwagon.ElementsToBytesUncompressed(commitments)
 	consumed := 0
 	for i := 0; i < NodeWidth; i++ {
 		if bit(bitlist[:], i) {
 			// if a child is present and is a hash, add its
 			// internal, serialized representation directly.
 			if bit(hashlist[:], i) {
-				children = append(children, n.children[i].(*HashedNode).commitment...)
+				bytesUncompressed := n.children[i].(*HashedNode).BytesUncompressed()
+				children = append(children, bytesUncompressed[:]...)
 			} else {
 				children = append(children, bytecomms[consumed][:]...)
 				consumed++
 			}
 		}
 	}
 
-	return append(append([]byte{internalRLPType}, bitlist[:]...), children...), nil
+	// Store in ret the serialized result
+	ret[nodeTypeOffset] = internalRLPType
+	copy(ret[internalBitlistOffset:], bitlist[:])
+	// Note that children were already appended in ret through the children slice.
```

Comment:
Since `children` was already written in `ret` (via `children` internal slice), we only set the node type, and copy the bitlist which is 32 bytes (quick).